scipy/sparse/_dok.py

```diff
@@ -274,19 +372,55 @@ def _mul_scalar(self, other):
         return new
 
     def _matmul_vector(self, other):
-        # matrix * vector
-        result = np.zeros(self.shape[0], dtype=upcast(self.dtype, other.dtype))
+        res_dtype = upcast(self.dtype, other.dtype)
+
+        # vector @ vector
+        if self.ndim == 1:
+            if issparse(other):
+                if other.format == "dok":
+                    shared_keys = self.keys() & other.keys()
+                else:
+                    o_coo = other.tocoo()
+                    shared_keys = self.keys() & o_coo.coords[0]
+                arr = np.array([self._dict[k] * other._dict[k] for k in shared_keys])
+                return arr.sum(dtype=res_dtype)
+            elif isdense(other):
+                arr = np.array([other[k] * v for k, v in self.items()])
+                return arr.sum(dtype=res_dtype)
+
+        # matrix @ vector
+        result = np.zeros(self.shape[0], dtype=res_dtype)
         for (i, j), v in self.items():
             result[i] += v * other[j]
         return result
 
     def _matmul_multivector(self, other):
-        # matrix * multivector
-        result_shape = (self.shape[0], other.shape[1])
         result_dtype = upcast(self.dtype, other.dtype)
-        result = np.zeros(result_shape, dtype=result_dtype)
-        for (i, j), v in self.items():
-            result[i,:] += v * other[j,:]
+        # vector @ multivector
+        if self.ndim == 1:
+            if other.ndim == 1:
+                result_shape = ()
```

Comment:
Return scalar?

Reply:
To follow array api, return a 0d array with the dot product as the only value.